templates/cluster-template-flatcar.yaml

```diff
@@ -0,0 +1,265 @@
+        - hostPath: /etc/kubernetes/azure.json
+          mountPath: /etc/kubernetes/azure.json
+          name: cloud-config
+          readOnly: true
+      etcd:
+        local:
+          dataDir: /var/lib/etcddisk/etcd
+          extraArgs:
+            quota-backend-bytes: "8589934592"
+    diskSetup:
+      filesystems: []
+      partitions: []
+    files:
+    - contentFrom:
+        secret:
+          key: control-plane-azure.json
+          name: ${CLUSTER_NAME}-control-plane-azure-json
+      owner: root:root
+      path: /etc/kubernetes/azure.json
+      permissions: "0644"
+    format: ignition
+    ignition:
+      containerLinuxConfig:
+        additionalConfig: |
+          systemd:
+            units:
+            - name: kubeadm.service
+              dropins:
+              - name: 10-flatcar.conf
+                contents: |
+                  [Unit]
+                  After=oem-cloudinit.service
+          # Workaround for https://github.com/kubernetes-sigs/cluster-api/issues/7679.
+          storage:
+            disks:
+            - device: /dev/disk/azure/scsi1/lun0
+              partitions:
+              - number: 1
+            filesystems:
+            - name: etcd_disk
+              mount:
+                device: /dev/disk/azure/scsi1/lun0
+                format: ext4
+                label: etcd_disk
+                options:
+                - -E
+                - lazy_itable_init=1,lazy_journal_init=1
+    initConfiguration:
+      nodeRegistration:
+        kubeletExtraArgs:
+          azure-container-registry-config: /etc/kubernetes/azure.json
+          cloud-config: /etc/kubernetes/azure.json
+          cloud-provider: azure
+        name: '@@HOSTNAME@@'
+    joinConfiguration:
+      nodeRegistration:
+        kubeletExtraArgs:
+          azure-container-registry-config: /etc/kubernetes/azure.json
+          cloud-config: /etc/kubernetes/azure.json
+          cloud-provider: azure
+        name: '@@HOSTNAME@@'
+    mounts:
+    - - etcd_disk
+      - /var/lib/etcddisk
+    postKubeadmCommands: []
+    preKubeadmCommands:
+    - sed -i "s/@@HOSTNAME@@/$(curl -s -H Metadata:true --noproxy '*' 'http://169.254.169.254/metadata/instance?api-version=2020-09-01'
+      | jq -r .compute.name)/g" /etc/kubeadm.yml
+  machineTemplate:
+    infrastructureRef:
+      apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+      kind: AzureMachineTemplate
+      name: ${CLUSTER_NAME}-control-plane
+  replicas: ${CONTROL_PLANE_MACHINE_COUNT}
+  version: ${KUBERNETES_VERSION}
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: AzureMachineTemplate
+metadata:
+  name: ${CLUSTER_NAME}-control-plane
+  namespace: default
+spec:
+  template:
+    spec:
+      dataDisks:
+      - diskSizeGB: 256
+        lun: 0
+        nameSuffix: etcddisk
+      image:
+        computeGallery:
+          gallery: flatcar4capi-742ef0cb-dcaa-4ecb-9cb0-bfd2e43dccc0
+          name: flatcar-stable-amd64-capi-${KUBERNETES_VERSION}
```

Comment:
will there be a new image published with each k8s version? if so, what is the expected delay betwen the k8s version being released and the image being available? Since our e2e test is wired to use the latest stable version available in the Azure VM image gallery, this could cause a race where the image is available for Ubuntu but not Flatcar yet and block PRs / fail periodic jobs.

Reply:
As mentioned in PR description, we plan to at least fulfill whatever is needed for CI here. Could you elaborate on the wiring? How are images published into Azure VM image gallery right now? Are they being published right after the new k8s release? I thought the k8s version used in e2e tests is periodically bumped by hand to be honest. CC @t-lo

Reply:
The E2E config specifies the k8s version as `latest-major.minor` https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/config/azure-dev.yaml#L151, which then gets resolve to the latest k8s version that has an available Windows _and_ Ubuntu image available: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/helpers.go#L713 in the Azure Marketplace.

I misspoke when I said VM image gallery earlier, I meant Azure Marketplace. We are planning on migrating the test image to a public Shared Image Gallery but currently, they are still in the Marketplace. The images are published by pipelines after each new k8s release but that process is still kicked off by hand (@mboersma can add more details on the publishing process if you have questions about that). 

We could change the logic to look for the latest Ubunut, Flatcar and Windows image available but that might slow down the rate at which we can test new versions if the Flatcar image publishing process is ad hoc / manual, which is why I asked.

Reply:
Probably @t-lo can add more to this and/or confirm, but if the requirement is to provide images for `latest-major.minor` K8s versions and building is still done by hand right now, I think Flatcar team can definitely commit to deliver those images. The building process is documented on our side already and we plan to automate it in the future.

Also, once all images used by CI are migrated to a public SIG, it shouldn't be a problem to build Flatcar images there as well then. We plan to upstream all required patches to image-builder and provide documentation on the building process.

Reply:
Thanks for the ping. We're currently aiming to integrate gallery publishing into our release process, and use the latest Kubernetes version at the time of a release. We usually release new images every 2-3 weeks. We'll start doing this manually but plan to eventually have it 100% automated (in fact, we're preferring a public image gallery over marketplace offers because publishing to a gallery is easily automatable).

While we don't have full automation in place it's hard to commit to immediate image releases tough when a new Kubernetes version becomes available since that would put additional load on the Flatcar OS maintainers team. This is transient, and because of our 2-3 week release cycle we'd be catching up soon enough with a new Kubernetes release. Is there a way to bypass running Flatcar tests in the CI when a respective image is not yet available?

Reply:
(also, CC: our release engineer @sayanchowdhury for visibility)

Reply:
> Is there a way to bypass running Flatcar tests in the CI when a respective image is not yet available

There is, but I'm not sure I agree that's something we should do since that could cause us to not have any signal on Flatcar for weeks without knowing it (test would show as "green" and I doubt people would check the output to see if the test ran). An alternative would be to add a separate k8s version for Flatcar and just have the Flatcar test run on its own latest available k8s version for that major.minor for Flatcar (we could add the func to discover what version that is in this PR).

@mboersma @jackfrancis @jsturtevant thoughts?

Reply:
100% agree w/ shipping a test that uses a different version of k8s for Flatcar compared to Ubuntu. This sort of makes sense, as our k8s version velocity at present is intrinsically related to the progress of Ubuntu VHD images that we maintain. Not surprising that another image publication pipeline would have a different k8s version velocity.

Reply:
Should I then implement a discovery for available K8s version for Flatcar images?

Reply:
I think that's right. You could define a new env var `KUBERNETES_VERSION_FLATCAR` with its own discovery implementation, and then just refer to that in the flatcar templates. That might be easier than overloading the existing `KUBERNETES_VERSION` in the existing discovery flow (we'd need to add a condition "is this for flatcar or no?", which might be more of a headache than simply discovering the latest flatcar-available version during each E2E run).

Thoughts @CecileRobertMichon @mboersma ?

Reply:
I've pushed initial implementation which I managed to get to work yesterday. I've tried to make code look similar to what's already there, although the structure seem rather arbitrary and I'm not sure where would I start to add tests for this code. But since this is an e2e test framework, I guess it's OK. Let me know what you think.

So now `FLATCAR_KUBERNETES_VERSION` is independent from `KUBERNETES_VERSION` in e2e tests, although templates still use `KUBERNETES_VERSION` as users would expect.
`FLATCAR_KUBERNETES_VERSION` supports `stable-1.24` format like `KUBERNETES_VERSION` and resolves to latest patch version by listing Flatcar community gallery images.
Once Kubernetes version is selected, we pick up latest Flatcar version for given Kubernetes version.

Reply:
This is now implemented and reviewed, so I'll mark as resolved.